File: modules/napplet-host/android/src/main/java/org/nostrocket/hypergolic/host/PublishedRelayWebSocketClientFrames.java
package org.nostrocket.hypergolic.host;

import java.io.ByteArrayOutputStream;
import java.nio.ByteBuffer;
import java.nio.charset.CharacterCodingException;
import java.nio.charset.CodingErrorAction;
import java.nio.charset.StandardCharsets;
import java.security.SecureRandom;

/** Bounded RFC 6455 client-to-server frame encoder; performs no I/O or fragmentation. */
public final class PublishedRelayWebSocketClientFrames {
  public static final int MAX_RAW_BYTES = 66_560;
  public static final int MAX_TEXT_BYTES = 8_192;
  private static final int MAX_CONTROL_BYTES = 125;
  private static final SecureRandom RANDOM = new SecureRandom();

  private PublishedRelayWebSocketClientFrames() {}

  /** Encodes one final UTF-8 text frame for the bounded relay lookup path. */
  public static byte[] textReq(String text) {
    return textReq(text, randomMask());
  }

  /** Deterministic proof overload. The mask is copied into the frame and must be exactly four bytes. */
  public static byte[] textReq(String text, byte[] mask) {
    if (text == null || text.length() > MAX_TEXT_BYTES) throw invalid();
    byte[] payload = strictUtf8(text);
    if (payload.length > MAX_TEXT_BYTES) throw invalid();
    return encode(0x1, payload, mask);
  }

  /** Encodes a pong that echoes the payload of a received ping. */
  public static byte[] pong(byte[] pingPayload) {
    return pong(pingPayload, randomMask());
  }

  /** Deterministic proof overload. */
  public static byte[] pong(byte[] pingPayload, byte[] mask) {
    if (pingPayload == null || pingPayload.length > MAX_CONTROL_BYTES) throw invalid();
    return encode(0xA, pingPayload.clone(), mask);
  }

  /** Encodes one close control frame with a status code and optional UTF-8 reason. */
  public static byte[] close(int code, String reason) {
    return close(code, reason, randomMask());
  }

  /** Deterministic proof overload. */
  public static byte[] close(int code, String reason, byte[] mask) {
    if (!validCloseCode(code) || reason == null || reason.length() > MAX_CONTROL_BYTES - 2) throw invalid();
    byte[] reasonBytes = strictUtf8(reason);
    if (reasonBytes.length > MAX_CONTROL_BYTES - 2) throw invalid();
    ByteArrayOutputStream payload = new ByteArrayOutputStream(2 + reasonBytes.length);
    payload.write((code >>> 8) & 0xff);
    payload.write(code & 0xff);
    payload.write(reasonBytes, 0, reasonBytes.length);
    return encode(0x8, payload.toByteArray(), mask);
  }

  private static byte[] encode(int opcode, byte[] payload, byte[] mask) {
    int lengthBytes = payload.length <= 125 ? 0 : 2;
    int payloadOffset = 2 + lengthBytes + 4;
    if (mask == null || mask.length != 4 || payload.length + payloadOffset > MAX_RAW_BYTES) throw invalid();
    if (payload.length > MAX_TEXT_BYTES || payload.length > MAX_CONTROL_BYTES && opcode >= 8) throw invalid();
    byte[] output = new byte[payload.length + payloadOffset];
    output[0] = (byte) (0x80 | opcode);
    output[1] = (byte) (0x80 | (lengthBytes == 0 ? payload.length : 126));
    if (lengthBytes != 0) {
      output[2] = (byte) ((payload.length >>> 8) & 0xff);
      output[3] = (byte) (payload.length & 0xff);
    }
    System.arraycopy(mask, 0, output, 2 + lengthBytes, 4);
    for (int i = 0; i < payload.length; i++) output[payloadOffset + i] = (byte) (payload[i] ^ mask[i & 3]);
    return output;
  }

  private static byte[] randomMask() {
    byte[] mask = new byte[4];
    RANDOM.nextBytes(mask);
    return mask;
  }

  private static byte[] strictUtf8(String value) {
    try {
      ByteBuffer encoded = StandardCharsets.UTF_8.newEncoder()
          .onMalformedInput(CodingErrorAction.REPORT)
          .onUnmappableCharacter(CodingErrorAction.REPORT)
          .encode(java.nio.CharBuffer.wrap(value));
      byte[] result = new byte[encoded.remaining()];
      encoded.get(result);
      return result;
    } catch (CharacterCodingException invalidUtf8) { throw invalid(); }
  }

  private static boolean validCloseCode(int code) {
    return code >= 1000 && code < 5000 && code != 1004 && code != 1005 && code != 1006
        && code != 1015 && !(code >= 1016 && code < 3000);
  }

  private static IllegalArgumentException invalid() {
    return new IllegalArgumentException("Invalid bounded WebSocket client frame");
  }
}

File: modules/napplet-host/ios/PublishedRelayWebSocketClientFrames.swift
import Foundation
import Security

/// Bounded RFC 6455 client-to-server frame encoder. It performs no I/O or fragmentation.
final class PublishedRelayWebSocketClientFrames {
    static let maxRawBytes = 66_560
    static let maxTextBytes = 8_192
    private static let maxControlBytes = 125

    /// Encodes one final UTF-8 text frame for the bounded relay lookup path.
    static func textReq(_ text: String) throws -> Data {
        try textReq(text, mask: secureMask())
    }

    /// Deterministic proof overload. The mask must be exactly four bytes.
    static func textReq(_ text: String, mask: Data) throws -> Data {
        guard text.utf8.count <= maxTextBytes else { throw invalid() }
        let payload = Data(text.utf8)
        guard payload.count <= maxTextBytes, String(data: payload, encoding: .utf8) == text else { throw invalid() }
        return try encode(opcode: 0x1, payload: payload, mask: mask)
    }

    /// Encodes a pong echoing a received ping payload.
    static func pong(_ pingPayload: Data) throws -> Data {
        try pong(pingPayload, mask: secureMask())
    }

    /// Deterministic proof overload.
    static func pong(_ pingPayload: Data, mask: Data) throws -> Data {
        guard pingPayload.count <= maxControlBytes else { throw invalid() }
        return try encode(opcode: 0xA, payload: pingPayload, mask: mask)
    }

    /// Encodes a close frame with a status code and optional UTF-8 reason.
    static func close(code: Int, reason: String) throws -> Data {
        try close(code: code, reason: reason, mask: secureMask())
    }

    /// Deterministic proof overload.
    static func close(code: Int, reason: String, mask: Data) throws -> Data {
        guard validCloseCode(code), reason.utf8.count <= maxControlBytes - 2 else { throw invalid() }
        let reasonData = Data(reason.utf8)
        guard reasonData.count <= maxControlBytes - 2,
              String(data: reasonData, encoding: .utf8) == reason else { throw invalid() }
        var payload = Data([UInt8((code >> 8) & 0xff), UInt8(code & 0xff)])
        payload.append(reasonData)
        return try encode(opcode: 0x8, payload: payload, mask: mask)
    }

    private static func encode(opcode: UInt8, payload: Data, mask: Data) throws -> Data {
        let lengthBytes = payload.count <= 125 ? 0 : 2
        let payloadOffset = 2 + lengthBytes + 4
        guard mask.count == 4, payload.count + payloadOffset <= maxRawBytes,
              payload.count <= maxTextBytes,
              opcode < 8 || payload.count <= maxControlBytes else { throw invalid() }
        var output = Data([0x80 | opcode, 0x80 | UInt8(lengthBytes == 0 ? payload.count : 126)])
        if lengthBytes != 0 {
            output.append(UInt8((payload.count >> 8) & 0xff))
            output.append(UInt8(payload.count & 0xff))
        }
        output.append(mask)
        for (index, byte) in payload.enumerated() { output.append(byte ^ mask[index & 3]) }
        return output
    }

    private static func secureMask() throws -> Data {
        var bytes = [UInt8](repeating: 0, count: 4)
        guard SecRandomCopyBytes(kSecRandomDefault, bytes.count, &bytes) == errSecSuccess else { throw invalid() }
        return Data(bytes)
    }

    private static func validCloseCode(_ code: Int) -> Bool {
        code >= 1000 && code < 5000 && ![1004, 1005, 1006, 1015].contains(code) && !(1016..<3000).contains(code)
    }

    private static func invalid() -> NSError {
        NSError(domain: "PublishedRelayWebSocketClientFrames", code: 1,
                userInfo: [NSLocalizedDescriptionKey: "Invalid bounded WebSocket client frame"])
    }
}
